### ISSUE TITLE
FLAC plugin's seek is not sample-accurrate

### ISSUE THREAD

**Issue by Wang-Yue** [cmus/cmus#678]
Perhaps other plugins share the same problem. but I just have chance to compare the wav and flac. This may also be the main cause of https://github.com/cmus/cmus/issues/600.

When playing a flac file from specific location(for instance 0:01, or even 0:00), the seeking location is wrong. 

In my test, when asked to play from 0:01, it should output starting from location 176400(which is 44100*2*2) of the raw pcm. but it actually start from location 202752, which missed a lot of samples. to fill the length sometimes redundant samples are copied. Which leads to the issue discussed in https://github.com/cmus/cmus/issues/600. 

Opened this new bug since other plugins are affected too.

**Comment by mahkoh** [comment]
Afaik most formats simply don't support this kind of operation.

**Comment by Wang-Yue** [comment]
@mahkoh right, mp3 doesn't  but no one saves entire cd image as mp3s.  but libflac does support sample accurate seek.  this is of vital importance to cue support. flac should have similar support compared to wav format's.

**Comment by mahkoh** [comment]
A patch implementing accurate seeking in the flac plugin would be accepted.

**Comment by Wang-Yue** [comment]
please apply the following patch:
```
diff --git a/ip/flac.c b/ip/flac.c
index 5a8399b..95a6f22 100644
--- a/ip/flac.c
+++ b/ip/flac.c
@@ -58,8 +58,6 @@ struct flac_private {
 	struct keyval *comments;
 	double duration;
 	long bitrate;
-
-	unsigned int ignore_next_write : 1;
 };
 
 static T(ReadStatus) read_cb(const Dec *dec, unsigned char *buf, size_t *size, void *data)
@@ -161,15 +159,6 @@ static FLAC__StreamDecoderWriteStatus write_cb(const Dec *dec, const FLAC__Frame
 	int ch, nch, i = 0;
 	char *dest; int32_t src;
 
-	if (ip_data->sf == 0) {
-		return FLAC__STREAM_DECODER_WRITE_STATUS_CONTINUE;
-	}
-
-	if (priv->ignore_next_write) {
-		priv->ignore_next_write = 0;
-		return FLAC__STREAM_DECODER_WRITE_STATUS_CONTINUE;
-	}
-
 	frames = frame->header.blocksize;
 	channels = sf_get_channels(ip_data->sf);
 	bits = sf_get_bits(ip_data->sf);
@@ -417,8 +406,7 @@ static int flac_read(struct input_plugin_data *ip_data, char *buffer, int count)
 }
 
 /* Flush the input and seek to an absolute sample. Decoding will resume at the
- * given sample. Note that because of this, the next write callback may contain
- * a partial block.
+ * given sample.
  */
 static int flac_seek(struct input_plugin_data *ip_data, double offset)
 {
@@ -426,12 +414,20 @@ static int flac_seek(struct input_plugin_data *ip_data, double offset)
 	uint64_t sample;
 
 	sample = (uint64_t)(offset * (double)sf_get_rate(ip_data->sf) + 0.5);
-	if (!F(seek_absolute)(priv->dec, sample)) {
-		return -IP_ERROR_ERRNO;
+        if (!F(seek_absolute)(priv->dec, sample)) {
+            return -IP_ERROR_ERRNO;
+        }
+	while (F(get_state(priv->dec)) == FLAC__STREAM_DECODER_SEEK_ERROR) {
+		if (!F(flush)(priv->dec)) {
+			d_print("failed to flush\n");
+			return -IP_ERROR_ERRNO;
+		}
+		priv->buf_rpos = 0;
+		priv->buf_wpos = 0;
+		if (!F(seek_absolute)(priv->dec, sample)) {
+			return -IP_ERROR_ERRNO;
+		}
 	}
-	priv->ignore_next_write = 1;
-	priv->buf_rpos = 0;
-	priv->buf_wpos = 0;
 	return 0;
 }
```

see https://xiph.org/flac/api/group__flac__stream__decoder.html#ga6a2eb6072b9fafefc3f80f1959805ccb for detail.

**Comment by Wang-Yue** [comment]
There's one other bug in cue.c. I'll post shortly.

**Comment by Wang-Yue** [comment]
done

**Comment by Wang-Yue** [comment]
fixed by #679
